modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -682,66 +661,98 @@ private static void validateDynamicParameters(int expectedParamsCount, Object[]
+         */
+        private Queue<ScriptStatement> prepareStatementsQueue(List<ParsedResult> parsedResults, Object[] params) {
+            assert !parsedResults.isEmpty();
+
+            int paramsCount = parsedResults.stream().mapToInt(ParsedResult::dynamicParamsCount).sum();
+
+            validateDynamicParameters(paramsCount, params);
+
+            ScriptStatement[] results = new ScriptStatement[parsedResults.size()];
+
+            // We fill parameters in reverse order, because each script statement
+            // requires a reference to the future of the next statement.
+            CompletableFuture<AsyncSqlCursor<InternalSqlRow>> prevCursorFuture = null;
+            for (int i = parsedResults.size() - 1; i >= 0; i--) {
+                ParsedResult result = parsedResults.get(i);
+
+                Object[] params0 = Arrays.copyOfRange(params, paramsCount - result.dynamicParamsCount(), paramsCount);
+                paramsCount -= result.dynamicParamsCount();
+
+                results[i] = new ScriptStatement(result, params0, prevCursorFuture);
+                prevCursorFuture = results[i].cursorFuture;
             }
 
-            ScriptStatementParameters parameters = statements.poll();
+            return new ArrayBlockingQueue<>(results.length, false, List.of(results));
+        }
+
+        CompletableFuture<AsyncSqlCursor<InternalSqlRow>> processNext() {
+            ScriptStatement scriptStatement = statements.poll();
 
-            assert parameters != null;
+            assert scriptStatement != null;
 
-            ParsedResult parsedResult = parameters.parsedResult;
-            Object[] dynamicParams = parameters.dynamicParams;
-            CompletableFuture<AsyncSqlCursor<InternalSqlRow>> cursorFuture = parameters.cursorFuture;
-            CompletableFuture<AsyncSqlCursor<InternalSqlRow>> nextCursorFuture = parameters.nextStatementFuture;
+            CompletableFuture<AsyncSqlCursor<InternalSqlRow>> cursorFuture = scriptStatement.cursorFuture;
 
             try {
                 if (cursorFuture.isDone()) {
                     return cursorFuture;
                 }
 
-                QueryTransactionWrapper txWrapper = wrapTxOrStartImplicit(parsedResult.queryType(), transactions, explicitTransaction);
+                ParsedResult parsedResult = scriptStatement.parsedResult;
 
-                QueryCancel cancel = new QueryCancel();
+                QueryTransactionWrapper txWrapper = scriptTxCtx.getOrStartImplicit(parsedResult.queryType());
 
-                executeParsedStatement(schemaName, parsedResult, txWrapper, cancel, dynamicParams, true, nextCursorFuture)
-                        .whenComplete((res, ex) -> {
+                scriptTxCtx.registerCursor(parsedResult.queryType(), cursorFuture);
+
+                executeStatement(parsedResult, txWrapper, scriptStatement.dynamicParams, scriptStatement.nextStatementFuture)
```

Comment:
If an implicit transaction was started above in 728, will control statement fails here?

Reply:
I don't quite understand the case you describe.

The implicit transaction at line 728 can only be started if this is NOT a control statement. Otherwise - even we fail on line 728 (nested tx) or we will start/use script-driven transaction.